From 51d65afd723b71855d8ede8446889a69e6a4aceb Mon Sep 17 00:00:00 2001
From: ajayk <ajaykemparaj@gmail.com>
Date: Tue, 18 Feb 2025 22:43:43 -0800
Subject: [PATCH] crypto: Always use openssl for obsolete cryptography

Always use openssl (via boring indirection), when requesting to use
obsolete cryptography. Right now the "boring.Supports*" checks are
incorrect, as they allow fallback to gocrypto, when in practice, other
implementations lie that they support this and then block them at
runtime.
---
 src/crypto/des/cipher.go | 4 ++--
 src/crypto/md5/md5.go    | 4 ++--
 src/crypto/rc4/rc4.go    | 6 +++---
 3 files changed, 7 insertions(+), 7 deletions(-)

diff --git a/src/crypto/des/cipher.go b/src/crypto/des/cipher.go
index 0d3c4f810c..7360e259f4 100644
--- a/src/crypto/des/cipher.go
+++ b/src/crypto/des/cipher.go
@@ -37,7 +37,7 @@ func NewCipher(key []byte) (cipher.Block, error) {
 	if len(key) != 8 {
 		return nil, KeySizeError(len(key))
 	}
-	if boring.Enabled && boring.SupportsDESCipher() {
+	if boring.Enabled {
 		return boring.NewDESCipher(key)
 	}
 
@@ -88,7 +88,7 @@ func NewTripleDESCipher(key []byte) (cipher.Block, error) {
 	if len(key) != 24 {
 		return nil, KeySizeError(len(key))
 	}
-	if boring.Enabled && boring.SupportsTripleDESCipher() {
+	if boring.Enabled {
 		return boring.NewTripleDESCipher(key)
 	}
 
diff --git a/src/crypto/md5/md5.go b/src/crypto/md5/md5.go
index f7aa6da36f..2d3555789f 100644
--- a/src/crypto/md5/md5.go
+++ b/src/crypto/md5/md5.go
@@ -105,7 +105,7 @@ func consumeUint32(b []byte) ([]byte, uint32) {
 // [encoding.BinaryUnmarshaler] to marshal and unmarshal the internal
 // state of the hash.
 func New() hash.Hash {
-	if boring.Enabled && boring.SupportsHash(crypto.MD5) {
+	if boring.Enabled {
 		return boring.NewMD5()
 	}
 	d := new(digest)
@@ -192,7 +192,7 @@ func (d *digest) checkSum() [Size]byte {
 
 // Sum returns the MD5 checksum of the data.
 func Sum(data []byte) [Size]byte {
-	if boring.Enabled && boring.SupportsHash(crypto.MD5) {
+	if boring.Enabled {
 		if fips140only.Enabled {
 			panic("crypto/md5: use of MD5 is not allowed in FIPS 140-only mode")
 		}
diff --git a/src/crypto/rc4/rc4.go b/src/crypto/rc4/rc4.go
index 09ebdfd733..beea0bdd47 100644
--- a/src/crypto/rc4/rc4.go
+++ b/src/crypto/rc4/rc4.go
@@ -41,7 +41,7 @@ func NewCipher(key []byte) (*Cipher, error) {
 	if k < 1 || k > 256 {
 		return nil, KeySizeError(k)
 	}
-	if boring.Enabled && boring.SupportsRC4() {
+	if boring.Enabled {
 		c, err := boring.NewRC4Cipher(key)
 		if err != nil {
 			return nil, err
@@ -65,7 +65,7 @@ func NewCipher(key []byte) (*Cipher, error) {
 // Deprecated: Reset can't guarantee that the key will be entirely removed from
 // the process's memory.
 func (c *Cipher) Reset() {
-	if boring.Enabled && boring.SupportsRC4() {
+	if boring.Enabled {
 		c.boring.Reset()
 		return
 	}
@@ -78,7 +78,7 @@ func (c *Cipher) Reset() {
 // XORKeyStream sets dst to the result of XORing src with the key stream.
 // Dst and src must overlap entirely or not at all.
 func (c *Cipher) XORKeyStream(dst, src []byte) {
-	if boring.Enabled && boring.SupportsRC4() {
+	if boring.Enabled {
 		c.boring.XORKeyStream(dst, src)
 		return
 	}
-- 
2.39.5 (Apple Git-154)

